// cards/build/decks/tarot.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TarotDeck = exports.TarotTrumpSuit = void 0;
const _deck_1 = require("./_deck");
const cards_1 = require("../cards");
const suits_1 = require("../suits");
const ranks_1 = require("../ranks");
var TarotTrumpSuit;
(function (TarotTrumpSuit) {
    TarotTrumpSuit["modern"] = "modern";
    TarotTrumpSuit["traditional"] = "traditional";
})(TarotTrumpSuit = exports.TarotTrumpSuit || (exports.TarotTrumpSuit = {}));
const suits = [suits_1.swords, suits_1.cups, suits_1.coins, suits_1.wands];
const ranks = [
    ranks_1.standard.ace,
    ranks_1.standard.nine,
    ranks_1.standard.ten,
    ranks_1.standard.jack,
    ranks_1.standard.queen,
    ranks_1.standard.king
];
const trumpRanks = [
    ranks_1.trump.trump0, ranks_1.trump.trump1, ranks_1.trump.trump2, ranks_1.trump.trump3, ranks_1.trump.trump4, ranks_1.trump.trump5, ranks_1.trump.trump6, ranks_1.trump.trump7,
    ranks_1.trump.trump8, ranks_1.trump.trump9, ranks_1.trump.trump10, ranks_1.trump.trump11, ranks_1.trump.trump12, ranks_1.trump.trump13, ranks_1.trump.trump14,
    ranks_1.trump.trump15, ranks_1.trump.trump16, ranks_1.trump.trump17, ranks_1.trump.trump18, ranks_1.trump.trump19, ranks_1.trump.trump20, ranks_1.trump.trump21
];
const arcanaRanks = [
    ranks_1.majorArcana.arcana0, ranks_1.majorArcana.arcana1, ranks_1.majorArcana.arcana2, ranks_1.majorArcana.arcana3, ranks_1.majorArcana.arcana4, ranks_1.majorArcana.arcana5, ranks_1.majorArcana.arcana6, ranks_1.majorArcana.arcana7,
    ranks_1.majorArcana.arcana8, ranks_1.majorArcana.arcana9, ranks_1.majorArcana.arcana10, ranks_1.majorArcana.arcana11, ranks_1.majorArcana.arcana12, ranks_1.majorArcana.arcana13, ranks_1.majorArcana.arcana14,
    ranks_1.majorArcana.arcana15, ranks_1.majorArcana.arcana16, ranks_1.majorArcana.arcana17, ranks_1.majorArcana.arcana18, ranks_1.majorArcana.arcana19, ranks_1.majorArcana.arcana20, ranks_1.majorArcana.arcana21
];
class TarotDeck extends _deck_1.Deck {
    constructor(opts = {}) {
        const cards = TarotDeck.generateCards(opts.trumpSuit || TarotTrumpSuit.modern);
        super({
            cards,
            rng: opts.rng
        });
    }
    static generateCards(trumpSuit) {
        const cards = [];
        // The minor arcana
        suits.forEach((suit) => {
            ranks.forEach((rank) => {
                cards.push(new cards_1.Card(suit, rank));
            });
        });
        // The major arcana
        if (trumpSuit === 'modern') {
            trumpRanks.forEach((rank) => {
                cards.push(new cards_1.Card(suits_1.trump, rank));
            });
        }
        else if (trumpSuit === 'traditional') {
            arcanaRanks.forEach((rank) => {
                cards.push(new cards_1.Card(suits_1.majorArcana, rank));
            });
        }
        return cards;
    }
}
exports.TarotDeck = TarotDeck;
